blog/2023-03-21-concurrency-in-rest-apis/index.mdx

```diff
@@ -0,0 +1,175 @@
+
+## What are ETags?
+
+ETag can be described as a fingerprint of a resource. It's a unique identifier that is generated based on the content of the resource. If the content changes, the ETag changes as well.
+ETag is used for caching purposes, as well as for concurrency control, which is our use case.
+
+### Weak vs Strong ETags
+
+ETags can be either **strong** or **weak**. Strong ETags guarantee byte-for-byte equality. They are more expensive to generate. Weak ETags only guarantee that the content is semantically equal.
+
+For our use case, we will use weak ETags.
+
+The format of a weak ETag is `W/"ETAG_HERE"`. The `W/` prefix indicates that the ETag is weak.
+
+<details>
+<summary><code>ETag</code> vs <code>Last-Modified</code></summary>
+
+`Last-Modified` header is another way to handle caching/concurrency. How does it work? When we fetch a resource from the server, we get a `Last-Modified` header in the response. This header contains the **date
+and time when the resource was last modified**. When we want to update the resource, we send the `If-Unmodified-Since` header[^2] in the request. The server will check if the resource was modified after the date and
+time specified in the header. If it was, the update will be rejected with `412 Precondition Failed` response. If it wasn't, the server will proceed with handling the request.
+
+The issue with this approach is its precision. The `Last-Modified` header is only precise to the second. If the resource is updated multiple times in the same second, the `If-Unmodified-Since` header will not be
+able to distinguish between the updates.
+
+</details>
+
+## Real-world example
+
+If we have no special handling of concurrent requests, this leads to **"lost updates"** problem. This means that the later request overwrites the changes made by the earlier requests, making them "lost".
+
+Let's demonstrate the problem of lost updates with an example. Let's, for a moment, pretend there's no such thing as Google Docs or similar collaborative editing tools.
+Our boss comes to us and says:
+
+> We need a tool that will allow group of people to **work together on a text document**. It needs to be finished in **2 days**.
+
+With all the time in the world, we came up with the following design:
+
+* This **will not be** a real-time collaboration where we can see the changes as they happen
+* A user will only be able to update the document if the document is based on the **latest version of the document**
+
+The business was happy with the design proposal, so we started implementing it.
+
+### The user experience
+
+After we finished the implementation, we tested it, by inviting 2 of our colleagues (Jeannie and Frank) to test it out.
+The following is a timeline of Jeanie and Frank's actions and the results:
+
+![Action sequence](./action-sequence.png)
```

Comment:
This is a great way to show the difference between using etags and not using them. Having a second timeline, without etags, to compare and understand the problem would be nice.